autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnLoadEvaluator.java

```diff
@@ -0,0 +1,180 @@
+                    () -> {
+                        yarnResponse.getScaleDownCandidates().ifPresent(
+                                scaleDownCandidates -> handleScaleDown(scaleDownCandidates, hostFqdnsToInstanceId));
+                    });
+
+            LOGGER.debug("Finished loadEvaluator for cluster {} in {} ms", cluster.getStackCrn(), System.currentTimeMillis() - start);
+        } catch (Exception ex) {
+            LOGGER.info("Failed to process load alert for Cluster {}, exception {}", cluster.getStackCrn(), ex);
+            eventPublisher.publishEvent(new UpdateFailedEvent(clusterId));
+        } finally {
+            LOGGER.debug("Finished loadEvaluator for cluster {} in {} ms", cluster.getStackCrn(), System.currentTimeMillis() - start);
+        }
+    }
+
+    protected void handleScaleUp(String hostGroupInstanceType, NewNodeManagerCandidates newNMCandidates, Integer existingHostGroupSize) {
+        Integer yarnRecommendedHostGroupCount =
+                newNMCandidates.getCandidates().stream()
+                        .filter(candidate -> candidate.getModelName().equalsIgnoreCase(hostGroupInstanceType))
+                        .findFirst().map(NewNodeManagerCandidates.Candidate::getCount)
+                        .orElseThrow(() ->
+                                new RuntimeException(String.format("Yarn Scaling API Response does not contain recommended host count " +
+                                                "for hostGroupInstanceType %s in Cluster %s, Yarn Response %s",
+                                        hostGroupInstanceType, cluster.getStackCrn(), newNMCandidates)));
+
+        int maxAllowedScaleUp = loadAlert.getLoadAlertConfiguration().getMaxResourceValue() - existingHostGroupSize;
+        maxAllowedScaleUp = maxAllowedScaleUp < 0 ? 0 : maxAllowedScaleUp;
+        int scaleUpCount = Math.min(maxAllowedScaleUp, yarnRecommendedHostGroupCount);
+
+        LOGGER.info("ScaleUp NodeCount {} for Cluster {}, HostGroup {]", scaleUpCount,
+                cluster.getStackCrn(), loadAlert.getScalingPolicy().getHostGroup());
+
+        if (scaleUpCount > 0) {
+            loadAlert.getScalingPolicy().setScalingAdjustment(scaleUpCount);
+            eventPublisher.publishEvent(new ScalingEvent(loadAlert));
+        }
+    }
+
+    protected void handleScaleDown(List<DecommissionCandidate> decommissionCandidates, Map<String, String> hostGroupFqdnsToInstanceId) {
+        Set<String> hostGroupFqdns = hostGroupFqdnsToInstanceId.keySet();
+        int maxAllowedScaleDown = hostGroupFqdns.size() - loadAlert.getLoadAlertConfiguration().getMinResourceValue();
+        maxAllowedScaleDown = maxAllowedScaleDown < 0 ? 0 : maxAllowedScaleDown;
+
+        List<String> decommissionHostGroupNodeIds = decommissionCandidates.stream()
+                .sorted(Comparator.comparingInt(DecommissionCandidate::getAmCount))
+                .map(DecommissionCandidate::getNodeId)
+                .map(nodeFqdn -> nodeFqdn.split(":")[0])
```

Comment:
How efficient / inefficient is this approach? i.e. will it end up iterating over the same lists multiple times over (including string split, etc) for different nodes which need to be removed?

Reply:
new streams are produced at result generation steps like filter, map and  there will not be observable perf difference unless the list size is in 10000s. also streams processing has become standard list processing mechanism.